src/Engine/Utils.java

```diff
@@ -143,30 +148,171 @@ private static boolean checkSpace(Unit unit, GameInstance myGame, SearchNode cur
 		return betterRoute;
 	}
 	
+	private static int findMoveCost(Unit unit, int x, int y, GameMap map)
+	{
+		return unit.model.propulsion.getMoveCost(map.getEnvironment(x, y));
+	}
+
+	public static boolean isPathValid(Unit unit, Path path, GameMap map)
+	{
+		//System.out.println("Checking path validity. Length: " + (path.getPathLength()-1));
+		boolean canReach = true;
+
+		// Make sure the first waypoint is under the Unit.
+		if(path.getPathLength() <= 0 || path.getWaypoint(0).x != unit.x || path.getWaypoint(0).y != unit.y)
+		{
+			canReach = false;
+		}
+
+		int movePower = Math.min(unit.model.movePower, unit.fuel);
+
+		// Index from 1 so we don't count the space the unit is on.
+		for(int i = 1; canReach && (i < path.getPathLength()); ++i)
+		{
+			//System.out.println("Moving over " + map.getEnvironment(path.getWaypoint(i).x, path.getWaypoint(i).y).terrainType);
+			movePower -= findMoveCost(unit, path.getWaypoint(i).x, path.getWaypoint(i).y, map);
+			if(movePower < 0)
+			{
+				canReach = false;
+			}
+		}
+
+		return canReach;
+	}
+
+	/**
+	 * Calculates and returns the shortest path for unit to take from its current location to map(x, y).
+	 * If no valid path is found, null is returned.
+	 */
+	public static Path findShortestPath(Unit unit, int x, int y, GameMap map)
```

Comment:
Correct me if I'm wrong, but this looks like it duplicates the stuff from the findPossibleDestinations function. I'm thinking it'd be good stuff to make a separate function that returns the 2D array of moves left.
It could easily be private, I imagine.